### PR TITLE
Backport PR #38120: API: preserve freq in DTI/TDI.factorize

### PR DESCRIPTION
**Pull request by simonjayhawkins**
Backport PR #38120

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
@jbrockmendel if you can give this a quick eyeball before I merge. several merge conflicts where I could have gone wrong. see also comments in #38137

**Comment by jbrockmendel** [comment]
taking a look now

File: pandas/core/arrays/period.py [hunk 1]
```diff
@@ -766,6 +767,9 @@ def _check_timedeltalike_freq_compat(self, other):
 
         raise raise_on_incompatible(self, other)
 
+    def factorize(self, na_sentinel=-1):
+        return ExtensionArray.factorize(self, na_sentinel=na_sentinel)
+
```
Comment (jbrockmendel) [inline pandas/core/arrays/period.py:772]: why is this necessary?

**Comment by jbrockmendel** [comment]
The PeriodArray thing looks unnecessary but harmless, otherwise LGTM

**Comment by simonjayhawkins** [comment]
> The PeriodArray thing looks unnecessary but harmless, otherwise LGTM

see https://github.com/pandas-dev/pandas/pull/38137#issuecomment-735395004

**Comment by jbrockmendel** [comment]
makes sense, thanks
